fastlane/fastlane
# [deliver] introduce `timeout` for screenshots processing waiting time

### Checklist


### Motivation and Context

Resolves #20619

I support two applications with 20+ languages and over 200 screenshots.

- It addresses an open issue https://github.com/fastlane/fastlane/issues/20619 which has gained significant community attention

- _fastlane_ doesn't have `timeout` on screenshots processing, so if some screenshots get stuck in a _processing_ state, which happens quite often when there are hundreds of screenshots, the process will be endless.
- Since _fastlane_ doesn't have processing time timeout, a user has several options to handle stuck screenshots:
  - remove screenshots by themselves directly in App Store Connect, then _fastlane_ will re-upload screenshots automatically
    - there's no info on which screenshot stuck - so the user ends up looking up all of the localizations and all the screenshots
    - if there are several multi-language applications that get regularly stuck, updating screenshots becomes a nightmare
    - it's not really an automated process
  - terminate _fastlane_ process and repeat the _deliver_
    - screenshots get stuck regularly when there are hundreds of them, so retrying the whole bunch doesn't eventually succeed
  - slowing down the upload (for example, by setting `DELIVER_NUMBER_OF_THREADS` to `1`) and hoping that App Store Connect will be able to process them
     - doesn't really work - screenshots stuck with the same regularity as with `DELIVER_NUMBER_OF_THREADS` equals to `10`

As a result, neither of the attempted solutions doesn't really work for multi-language applications, so we need to introduce some workaround

### Description

The proposed solution is based on existing logic, which removes & re-uploads screenshots that failed to upload.
The `screenshot_processing_timeout` parameter is introduced within the `deliver` action. This parameter allows users to define a timeout period after which screenshots that are still processing are considered stuck (failed) and then existing logic with removing & reuploading comes into play.

- Added `screenshot_processing_timeout` parameter to `deliver` which defines the timeout of waiting before considering currently processing screenshots as failed ones.
- Screenshots that considered as failed removed from App Store Connect, then uploading is repeated. By doing so, one retry is used
- As a result, from the user perspective now it's possible to specify a reasonable timeout after which _deliver_ attempts to re-upload processing screenshots

Chosen constants:
- `1hr` is default value for `screenshot_processing_timeout`. It is chosen with backward compatibility in mind. I assume for anyone who uses `deliver`, `1hr` should be sufficient enough to not break existing logic

### Testing Steps

To test implementation, you should add hundreds of screenshots to App Store Connect with _deliver_. In my case, I have about 20 localizations with 10-12 screenshots in each (5.5 and 6.5 screens)
I've tested with 2 minutes for `screenshot_processing_timeout`. In my experience, it's usually enough for spotting stuck screenshots inside the EU. However, `screenshot_processing_timeout` may depend on your location, bandwidth, etc
In my experience I've encountered 2-3 per ~200 screenshots, so it's about 1%

----

To test this branch, modify your Gemfile as:

```ruby
gem 'fastlane', git: 'https://github.com/mikhailmaslo/fastlane.git', branch: 'feature/introduce-timeout-for-screenshots-upload'
```

And run `bundle install` to apply the changes.
